dolosse/generic_json_consumer.py

```diff
@@ -0,0 +1,153 @@
+                    variables[name].append(following)
+                else:
+                    variables[name].append(
+                        variables[name][-1])
+            else:
+                variables[name] = [0]*(count-1)
+                if valid:
+                    variables[name].append(following)
+                else:
+                    variables[name].append(0)
+                    #Remember, the list might still be empty here
+
+
+def get_args():
+    """Creates and returns a kafka consumer"""
+    parser = ArgumentParser()
+    parser.add_argument("--topic", default="default",
+                        help="The name of the kafka topic from which to read")
+    parser.add_argument("--group", default="my-group",
+                        help="The kafka group in which the topic exists")
+    parser.add_argument("--server", default="localhost:9092",
+                        help="Name and port of the kafka server (e.g. localhost:9092)")
+    arguments = parser.parse_args()
+
+    # To consume latest messages and auto-commit offsets
+    return arguments.topic, arguments.group, arguments.server
+
+def redraw(figures):
+    """Redraw all figures from a dict of figures"""
+    for key in figures:
+        figures[key].canvas.draw()
+        figures[key].canvas.flush_events()
+
+
+def update_data(data, value):
+    """Update the list of data, given some json input"""
+    try:
+        update = loads(value)
+    except decoder.JSONDecodeError as problem:
+        print("Error: Flawed JSON in value: ",
+              value)
+        print("Error message: ", problem)
+        update = ""
+    data.update(update)
+
+
+def graph(topic, group, server):
+    """Main function - reads json data from a kafka topic and graphs it"""
+
+    screen = initscr()
+    cbreak()
+    screen.nodelay(True)
+    screen.getch()
+
+    print("Generic Graphing Consumer")
+    print("Press any key to exit...")
+
+    data = {}
+    line = {}
+    plots = {}
+    figures = {}
+    axes = {}
+    count = 0
+    pyplot.ion() #Interactive mode on
+    configuration = {"bootstrap.servers": server,
+                     "group.id": group,
+                     "default.topic.config" : {
+                         "auto.offset.reset": "earliest"},
+                     "enable.auto.commit": False}
+    consumer = Consumer(configuration)
+    consumer.subscribe([topic])
+    keyboard_input = ERR
+
+    while True:
+        if keyboard_input != ERR:
+            break
+        keyboard_input = screen.getch()
+        message = consumer.consume(num_messages=1, timeout=0.2)
+        if (message is None) or (message == []):
+            redraw(figures)
+        elif message[0].error():
+            print("Something went wrong with the read!")
+            print(message.error())
+        else:
+            update_data(data, message[0].value())
+            count += 1
+
+            populate(line, data, count)
+
+            for key in line:
```

Comment:
I had to go back up to find out that `line` is a dictionary and not a string. I recommend changing the name of the variable to be clearer. Also, recommended way to loop over dictionary keys is `for key in line.keys():`.

Reply:
Hmmmm. It's a dictionary that contains all of the lines for the graphs.

...at the time when I named it, I wasn't thinking of a line of text. I'll try to think up a new name.

Reply:
Variable has been renamed.